[25.0] Fix inconsistent styling in List Collection Builder selector

ahmedhamidawan opened the pull request:
Fixes https://github.com/galaxyproject/galaxy/issues/20441:

<details>
<summary><i>Checklist from ticket ☑️ </i></summary>

- [x] Irregular alignment within dataset items, Aa, .* buttons, etc.
- [x] Buttons expand/contract on hover "click to select" behavior
- [x] Layout shifts causing repositioning during hover state
- [x] Are dataset names intentionally editable here? This mixed with shift to select led me to rename something randomly without an easy way to undo it.
</details>

| Before these changes 🐞 |
| ---- |
| <video src="https://github.com/user-attachments/assets/96e807d0-f78a-41ca-afda-df48e093a232" />
| Note that the styling on hover is not ideal, and renaming an element and selecting it - both being click operations can cause confusion and unintended consequences. Also a bug when you rename an element in the selected column, it gets moved back to unselected, however, the selected column still has a count of 2. |


| After these changes ✅ _(outdated, see screencasts in comments)_ |
| ---- |
| <video src="https://github.com/user-attachments/assets/a26c86d8-3312-4c15-8cdd-8cd1a6b6b6df" />
| Using the default view for our multiselects, with just text values as opposed to bordered boxes, helps fix hover/focus styling. Not renaming from the multiselect and instead using a separate modal fixes the rename issue as well. The last bug with rename causing elements to move around the selected/unselected lists unintentionally is also fixed. _At the end I demonstrate that this works nicely with a big history as well, as far as styling is concerned._ |

## How to test the changes?
(Select all options that apply)
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [x] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

ahmedhamidawan committed:
fix irregular alignment for the `FormSelectMany` search buttons

ahmedhamidawan committed:
use text only collection element; fix rename bug; add rename modal

This aims to fix all of the issues here: Fixes https://github.com/galaxyproject/galaxy/issues/20441

Using the default text-only value for the slot in `FormSelectMany` works best as it prevents layout shifting on hover/focused states.

This commit also fixes a bug with the `renameElement` function, which would create a duplicate of the renamed element (1 in `inListElements`, 1 in `workingElements` upon rename).

This adds a modal where the user can choose to rename elements if they'd like, instead of renaming the elements inside the multi select; which is not intuitive and can also interfere with clicking an element to select it.

dannon commented:
@ahmedhamidawan Nice, much improved!  Minor, but did you consider any ideas for relocating the warning message to avoid the page shift down?

This one:

![image](https://github.com/user-attachments/assets/1706acec-dac0-4a66-87d1-2af29e91c075)

ahmedhamidawan commented:
> @ahmedhamidawan Nice, much improved!  Minor, but did you consider any ideas for relocating the warning message to avoid the page shift down?
> 
> This one:
> 
> ![image](https://github.com/user-attachments/assets/1706acec-dac0-4a66-87d1-2af29e91c075)
> 

Thanks for taking a look!

This is from when this used to be in a modal, as opposed to in the run form.

I'm not sure where we could emit and track this, or move around in the creator itself 🤔

ahmedhamidawan committed:
unrelated to this PR; remove html tags from message which is rendered as is

We changed `FormError` to not use `v-html` in commit 1b01f86cafff39e88864821bbcb9d9d51639420d which broke this.

jmchilton commented:
>  Are dataset names intentionally editable here? 

What you are calling a dataset name is definitely intentionally editable - it just isn't the dataset name. You're choosing an element identifier for the collection element. It will persist through an analysis so it should generally not include extensions, funky characters, HIDs, etc... 

The language around "click a dataset to rename them" I do not like for this reason - it should be label the element identifier for the list - but maybe other people find it fine. I do see those being referred to as names sometimes - it just isn't the same at all to me.

The order of elements in a collection is really important - again it is persisted and included in reports and such when data reductions steps are done. My biggest issue if I recall correctly was that it didn't have a way to sort the elements. It really feels like you're selecting some datasets for a random operation like batch delete - not choosing some elements to construct into a list and then sorting and labelling the elements. Even if we think I'm being... esoteric... and we should keep the "rename" verb and such - we should at least have more help text available with popovers or clicks and such about what is going on and the consequences  of the "names" and the order of the elements.

ahmedhamidawan commented:
@jmchilton 

Thanks for taking a look!

Yes, you're right, it should say identifier, and i can remove the extra stuff like hid, etc from the rename modal because that is the stage when you're essentially looking at the final list. And I can even add sorting the list inside of that modal.

ahmedhamidawan committed:
change "dataset name" to "element identifier"; only show identifiers

(no `hid` or `extension`)

ahmedhamidawan commented:
What do you think @jmchilton ? Changed it to identifier, and also allowed reordering:

https://github.com/user-attachments/assets/5801ebf1-6d5d-4bce-a934-9ab60693bf4d

The help text can definitely be improved...

ahmedhamidawan committed:
allow list elements to be reordered as well

mvdbeek commented:
> And there were requests above (@jmchilton) to specifically not include the hids and extensions when you are in it, because at this stage these are simply identifiers for what will appear in the final collection.

how will you know what is what ? what John said is 

> It will persist through an analysis so it should generally not include extensions, funky characters, HIDs, etc...

and of course that makes sense, it shouldn't be part of the element identifier, but I don't think that's a request to only see the history dataset name.

I like the interface as it is now, the only thing missing was sort and rename, an edit button that switches those modes but otherwise uses the same components would seem like less of an intrusive action ?

ahmedhamidawan commented:
I guess one could make that change to the FormSelectMany multiselect, and allow reordering in there. I won't want to use the old styling again for the items (before these changes) as they appear in the multiselect, to allow for renaming.

Would it be ok to not be able to rename the elements for now?

ahmedhamidawan commented:
> renaming in the left side works fine), just gotta add them back in the right order....

Oh that I've actually fixed here by fixing the renameElement function in the 2nd commit

On modals, I reckon they're useful for absolutely optional actions like this one

ahmedhamidawan committed:
show the hid and extension (if applicable) to the rename modal elements

ahmedhamidawan commented:
What do we think of this for now as a compromise?:

![firefox_Mn0z0KOj3K](https://github.com/user-attachments/assets/2f078950-a5a9-4dc4-9cd2-188d18152fc7)

Rename files still inside of the multiselect, but since the styling is improved now, we have less interference from the click to edit component. And also, you can only click and edit to rename items already added to the list

Doesn't even need a toggle to switch to editing mode, since now its made even more evident that you can rename by clicking (with the column header text)

mvdbeek commented:
Nice, I like it!

ahmedhamidawan commented:
If we really want that, I have changes in this commit I can open a PR for. Otherwise, we can keep the modal

mvdbeek commented:
yes, please. the modal hides what you're doing, so not sure that the changes you're currently making at peristed (since there's no ok button), and the button to open the modal only appears after you already place things into the right side, which is again surprising, it shouldn't matter when I decide to edit the elements I think.
